### ISSUE TITLE
Refactor existing functions

### ISSUE THREAD

**Issue by strk** [NibioOpenSource/pgtopo_update_sql#29]
As per phone call: the plan is to replace `create_surface_edge_domain_obj` usage with `update_features` to prove it's working.
We'll focus on line 18 of src/test/sql/regress/topo_update.sql which goes:
> SELECT '3', count(id) FROM (SELECT 1 AS id FROM topo_update.create_surface_edge_domain_obj('{"type": "Feature","geometry":{"type":"LineString","crs":{"type":"name","properties":{"name":"EPSG:4258"}},"coordinates":[[18.3342803675,69.1937360885],[18.3248972004,69.1926352514],[18.3225223088,69.1928235904],[18.3172506318,69.1941599626],[18.3145519815,69.1957316656],[18.3123602886,69.1980059858],[18.310704822,69.2011722899],[18.3080083628,69.2036461481],[18.3052533657,69.2074983075],[18.3057756447,69.2082956989],[18.3075330509,69.2093067972],[18.3103134457,69.2100132313],[18.3156403748,69.2107656476],[18.3228118186,69.2113399389],[18.3301412606,69.2111984614],[18.3349532259,69.2112004097],[18.3395639862,69.2116335442],[18.3433515794,69.2127948707],[18.3502828982,69.2152724054],[18.3524811669,69.2156570867],[18.3547763375,69.2158024362],[18.3580354423,69.2152640418],[18.3623692173,69.2138971761],[18.3678152295,69.2110837518],[18.3695071064,69.2082009883],[18.3680734909,69.2067092134],[18.3638755844,69.2028967661],[18.355530639,69.1981677188],[18.3471464882,69.1957662158],[18.3342803675,69.1937360885]]},"properties":{"fellesegenskaper.opphav":null,"fellesegenskaper.kvalitet.synbarhet":null,"fellesegenskaper.kvalitet.noyaktighet":null,"fellesegenskaper.kvalitet.maalemetode":null,"fellesegenskaper.oppdateringsdato":null,"fellesegenskaper.forstedatafangstdato":"2016-01-01","fellesegenskaper.verifiseringsdato":null,"Fellesegenskaper.Kvalitet.Maalemetode":82}}','topo_rein', 'arstidsbeite_var_flate', 'omrade', 'arstidsbeite_var_grense','grense',  1e-10)) AS R;

In order to replace it with a call to `update_features` a couple of JSON mappings will be needed, one for the lineal layer and one for the areal layer.

**Comment by strk** [comment]
What version of GeoJSON is that ? Because according to https://geojson.org/ and https://tools.ietf.org/html/rfc7946 the "crs" member of the "geometry" field was removed, see https://tools.ietf.org/html/rfc7946#appendix-B.1

**Comment by larsop** [comment]
Yes that's correct today this is hard codet.  We also have the about issue setting default dates for some columns, this also this functin  

> As per phone call: the plan is to replace `create_surface_edge_domain_obj` usage with `update_features` to prove it's working.
> We'll focus on line 18 of src/test/sql/regress/topo_update.sql which goes:
> 
> > SELECT '3', count(id) FROM (SELECT 1 AS id FROM topo_update.create_surface_edge_domain_obj('{"type": "Feature","geometry":{"type":"LineString","crs":{"type":"name","properties":{"name":"EPSG:4258"}},"coordinates":[[18.3342803675,69.1937360885],[18.3248972004,69.1926352514],[18.3225223088,69.1928235904],[18.3172506318,69.1941599626],[18.3145519815,69.1957316656],[18.3123602886,69.1980059858],[18.310704822,69.2011722899],[18.3080083628,69.2036461481],[18.3052533657,69.2074983075],[18.3057756447,69.2082956989],[18.3075330509,69.2093067972],[18.3103134457,69.2100132313],[18.3156403748,69.2107656476],[18.3228118186,69.2113399389],[18.3301412606,69.2111984614],[18.3349532259,69.2112004097],[18.3395639862,69.2116335442],[18.3433515794,69.2127948707],[18.3502828982,69.2152724054],[18.3524811669,69.2156570867],[18.3547763375,69.2158024362],[18.3580354423,69.2152640418],[18.3623692173,69.2138971761],[18.3678152295,69.2110837518],[18.3695071064,69.2082009883],[18.3680734909,69.2067092134],[18.3638755844,69.2028967661],[18.355530639,69.1981677188],[18.3471464882,69.1957662158],[18.3342803675,69.1937360885]]},"properties":{"fellesegenskaper.opphav":null,"fellesegenskaper.kvalitet.synbarhet":null,"fellesegenskaper.kvalitet.noyaktighet":null,"fellesegenskaper.kvalitet.maalemetode":null,"fellesegenskaper.oppdateringsdato":null,"fellesegenskaper.forstedatafangstdato":"2016-01-01","fellesegenskaper.verifiseringsdato":null,"Fellesegenskaper.Kvalitet.Maalemetode":82}}','topo_rein', 'arstidsbeite_var_flate', 'omrade', 'arstidsbeite_var_grense','grense',  1e-10)) AS R;
> 
> In order to replace it with a call to `update_features` a couple of JSON mappings will be needed, one for the lineal layer and one for the areal layer.

Yes that's correct today this is hard codet.  

We also have  an issue about setting default dates for some columns, this is also this function today. You suggest trigger there and I think that's a good idea.

**Comment by strk** [comment]
I filed #30 to improve the `insert_feature` function (documentation and optional SRID).
What do we do with the CRS attribute in incoming JSON, do we want to honour it if present ?

**Comment by larsop** [comment]
> What version of GeoJSON is that ? Because according to https://geojson.org/ and https://tools.ietf.org/html/rfc7946 the "crs" member of the "geometry" field was removed, see https://tools.ietf.org/html/rfc7946#appendix-B.1

The crs has been a problem/discusion in GeoJSON since we started.  I am not sure if we are using it here. If it does not cause any problems just ignore this attribute or remove it from the test.

**Comment by strk** [comment]
I'm finding also nested composite types in PostgreSQL. This is another concern :(

The table `topo_rein.arstidsbeite_var_grense` has the `felles_egenskaper` attribute being of type `topo_rein.sosi_felles_egenskaper`, but then type `topo_rein.sosi_felles_egenskaper` has an attribute `kvalitet` which is another composite type `topo_rein.sosi_kvalitet` and also `registreringsversjon` being of type `topo_rein.sosi_registreringsversjon`. 

Is the current function populating them as well ? I don't think my current code supports nested composite types

**Comment by larsop** [comment]
> I'm finding also nested composite types in PostgreSQL. This is another concern :(
> 
> The table `topo_rein.arstidsbeite_var_grense` has the `felles_egenskaper` attribute being of type `topo_rein.sosi_felles_egenskaper`, but then type `topo_rein.sosi_felles_egenskaper` has an attribute `kvalitet` which is another composite type `topo_rein.sosi_kvalitet` and also `registreringsversjon` being of type `topo_rein.sosi_registreringsversjon`.
> 
> Is the current function populating them as well ? I don't think my current code supports nested composite types

Yes it is.  

The nesting is caused by trying to follow the spec here https://register.geonorge.no/data/documents/Produktspesifikasjoner_FKB%20AR5_v2_produktspesifikasjon-fkb-ar5-4_6_.pdf (sorry Norwegian)

**Comment by strk** [comment]
Ok then, filed #31 to also support nested complex types

**Comment by strk** [comment]
I'm on this task again. My current problem is with the test on line 19, which fetches rows from the table where...
> (felles_egenskaper).oppdateringsdato = current_date;

2 things:

  - 1: oppdateringsdato does not have a default value, in topo_rein.sosi_felles_egenskaper, should I add one ?
  - 2: the test seems suffering from a race condition: what if too much time passes from execution of line 18 and execution of line 19 ?

**Comment by larsop** [comment]
> I'm on this task again. My current problem is with the test on line 19, which fetches rows from the table where...
> 
> > (felles_egenskaper).oppdateringsdato = current_date;
> 
> 2 things:
> 
>     * 1: oppdateringsdato does not have a default value, in topo_rein.sosi_felles_egenskaper, should I add one ?

Yes (that picked up from res.oppdateringsdato :=  current_date; from function_01_get_felles_egenskaper.sql)
We may  move that to triggers as we discussed ?
> 
>     * 2: the test seems suffering from a race condition: what if too much time passes from execution of line 18 and execution of line 19 ?

Here I need some more info.

**Comment by strk** [comment]
I renamed the title of this ticket to reflect new plan, which is refactor the existing `create_surface_edge_domain_obj`

**Comment by strk** [comment]
What the function does, as understood from phone call is:

  - Receive a LINEAR feature ad a set of associated attributes as input (in JSON form)
  - Receive references to 2  TopoGeometry layers (a "surface" one and a "border" one, both sharing the same topology)
  - Do NOTHING if the LINEAR feature does not form NEW faces in the topology [ EARLY EXIT ! ]
  - Bail out if the LINEAR feature forms (splits) MORE THAN ONE face [ EARLY EXIT ! ]
  - Cut the line so it doesn't create any DANGLING edges in the topology
  - Insert the new face-splitting non-dangling line in the topology
  - Insert a new record in the "border" layer with a TopoGeometry covering the space of the non-dangling line and the attributes received in input
  - Update the "surface" layer so that any existing feature covering the face being split (ONLY ONE SUCH RECORD EXPECTED) is also split in two records, having the SAME attributes but TopoGeometry objects each being defined by each of the 2 faces resulting from the split

I think the above describes what the function does.

One question with this is about the last step (handling of "surface" layer): what identifier do we assign to the _new_ record being a duplication (as attributes) of the existing record ?

**Comment by larsop** [comment]
> What the function does, as understood from phone call is:
> 
>     * Receive a LINEAR feature ad a set of associated attributes as input (in JSON form)
> 
>     * Receive references to 2  TopoGeometry layers (a "surface" one and a "border" one, both sharing the same topology)
> 
>     * Do NOTHING if the LINEAR feature does not form NEW faces in the topology [ EARLY EXIT ! ]
> 
>     * Bail out if the LINEAR feature forms (splits) MORE THAN ONE face [ EARLY EXIT ! ]
> 
>     * Cut the line so it doesn't create any DANGLING edges in the topology
> 
>     * Insert the new face-splitting non-dangling line in the topology
> 
>     * Insert a new record in the "border" layer with a TopoGeometry covering the space of the non-dangling line and the attributes received in input
> 
>     * Update the "surface" layer so that any existing feature covering the face being split (ONLY ONE SUCH RECORD EXPECTED) is also split in two records, having the SAME attributes but TopoGeometry objects each being defined by each of the 2 faces resulting from the split
> 
> 
> I think the above describes what the function does.
> 
> One question with this is about the last step (handling of "surface" layer): what identifier do we assign to the _new_ record being a duplication (as attributes) of the existing record ?

We duplicate every thing except the primary key. Today the primary key is 'id' and is's serial. 
This is problem for ar5 which is uuid and not serial and is not named 'id'.

The missing parts now related to AR5
- Handle non serial primary keys with another name than 'id', maybe multiple columns 
- Handle default dates differently from layer to layer, trigger functions or mapping rules ?
- Generic mapping from JSON to table source based on mapping rules 
- The columns reinbeitebruker_id is hardcodet and the column name should a parameter. Today this is used to regulate access. 
-

**Comment by strk** [comment]
Is the rest of the function description correct ? Do we want the line to always only split ONE face ? Because I'm not seeing that check in the current `create_surface_edge_domain_obj`

**Comment by strk** [comment]
I've merged the first round of refactoring. Will continue after merging the schema loader script in PR #26
